src/gt4py/next/iterator/ir_utils/domain_utils.py

```diff
@@ -164,22 +166,106 @@ def translate(
-        )
-        stop = functools.reduce(
-            lambda current_expr, el_expr: im.maximum(current_expr, el_expr),
-            [domain.ranges[dim].stop for domain in domains],
-        )
-        # constant fold expression to keep the tree small
-        start, stop = ConstantFolding.apply(start), ConstantFolding.apply(stop)  # type: ignore[assignment]  # always an itir.Expr
-        new_domain_ranges[dim] = SymbolicRange(start, stop)
+        new_domain_ranges[dim] = range_reduce_op(*[domain.ranges[dim] for domain in domains])
 
     return SymbolicDomain(domains[0].grid_type, new_domain_ranges)
+
+
+domain_union = functools.partial(_reduce_domains, range_reduce_op=_range_union)
+"""Return the (set) union of a list of domains."""
+domain_intersection = functools.partial(_reduce_domains, range_reduce_op=_range_intersection)
+"""Return the intersection of a list of domains."""
+
+
+def domain_complement(domain: SymbolicDomain) -> SymbolicDomain:
+    """
+    Return the (set) complement of a half-infinite domain.
+
+    Note: after canonicalization of concat_where, the domain is always half-infinite,
+    i.e. it has ranges of the form `]-inf, a[` or `[a, inf[`.
+    """
+    dims_dict = {}
+    for dim in domain.ranges.keys():
+        lb, ub = domain.ranges[dim].start, domain.ranges[dim].stop
+        assert (lb == itir.InfinityLiteral.NEGATIVE) != (ub == itir.InfinityLiteral.POSITIVE)
+        # `]-inf, a[` -> `[a, inf[`
+        if lb == itir.InfinityLiteral.NEGATIVE:
+            dims_dict[dim] = SymbolicRange(start=ub, stop=itir.InfinityLiteral.POSITIVE)
+        # `[a, inf]` -> `]-inf, a]`
+        else:  # ub == itir.InfinityLiteral.POSITIVE:
+            dims_dict[dim] = SymbolicRange(start=itir.InfinityLiteral.NEGATIVE, stop=lb)
+    return SymbolicDomain(domain.grid_type, dims_dict)
+
+
+def promote_domain(
+    domain: SymbolicDomain, target_dims: Iterable[common.Dimension]
+) -> SymbolicDomain:
+    """Return a domain that is extended with the dimensions of target_dims."""
+    assert set(domain.ranges.keys()).issubset(target_dims)
+    dims_dict = {}
+    for dim in target_dims:
+        dims_dict[dim] = (
+            domain.ranges[dim]
+            if dim in domain.ranges
+            else SymbolicRange(itir.InfinityLiteral.NEGATIVE, itir.InfinityLiteral.POSITIVE)
+        )
+    return SymbolicDomain(domain.grid_type, dims_dict)
+
+
+def is_finite(range_or_domain: SymbolicRange | SymbolicDomain) -> bool:
+    """
+    Return whether a range is unbounded in (at least) one direction.
+
+    The expression is required to be constant folded before for the result to be reliable.
+    """
+    match range_or_domain:
+        case SymbolicRange() as range_:
+            if any(
+                v in [itir.InfinityLiteral.POSITIVE, itir.InfinityLiteral.NEGATIVE]
+                for v in [range_.start, range_.stop]
+            ):
+                return False
+            return True
```

Comment:
Isn't it simpler to write it as a single expression?
```suggestion
infinity_literals = {itir.InfinityLiteral.POSITIVE, itir.InfinityLiteral.NEGATIVE}
return not (range_.start in infinity_literals or range_.stop in infinity_literals)
```